blog/a-bloop-tour-for-metals-users.md

```diff
@@ -0,0 +1,141 @@
+protocol](https://github.com/facebook/nailgun). They can be happening
+concurrently, they cache compilations across different clients, and offer client
+isolation to avoid conflicts in a shared, stateful file system.
+
+The [_What is Bloop_](https://scalacenter.github.io/bloop/docs/what-is-bloop)
+portion of the website does a great job at outlining in more detail what I have
+listed above, along with some of the design goals of the project.
+
+## Metals and Bloop
+
+If you've seen any of the presentations done about Metals then you'll probably
+have seen a graphic similar to what I have below. However, the one below is a
+bit more simplified showing only one language client and one build tool.
+
+![BSP + LSP diagram](/media/diagram.png)
+
+This diagram is meant to show Bloops part in the flow from build definition to
+your editor of choice. If you start on the left, you have your build definition
+in your build tool of choice. We'll use SBT since it's the most common. If
+you've used Metals before, you'll notice that when you first open up a project,
+you'll be prompted to _Import Your Build_. What does this mean? The first time
+this happens, it means that Metals detects that you have no `.bloop` directory,
+and therefore you need to import your build. When using SBT this means that
+Metals actually adds the `sbt-bloop` plugin to your build in
+`project/metals.sbt`. It then issues an `sbt bloopInstall` command which will
+dump out your build definition for all of your modules into json files, which
+you can see if you open up your `.bloop` directory in your workspace. These
+files contain things like your directory information, what dependencies your
+module has on other modules, all your classpath information, and more. Go ahead
+and take a look at everything in there. Once this information is gathered,
+Metals tells Bloop to compile your project.
+
+At this point, if there are any errors during compilation, diagnostics are
+forwarded from Bloop to Metals, to your client for you to see. If you then fix
+the diagnostic, hit save, the save event is sent to Metals, which then forwards
+that to Bloop to compile what has been changed. You can start to imagine the
+flow from the diagram above.
+
+At this point, one of the questions you may have is how does Bloop start? Does
+Metals start it? Does Metals install Bloop? There can only be one Bloop server
+running on a machine. So when Metals is about to start, one of two things
+happens. Either Metals detects that a Bloop server is running and connects to
+it, or it will start one.
+
+If you've even used the debugging features in Metals, you're also utilizing the
+DAP (Debug Adaptor
+Protocol)[https://microsoft.github.io/debug-adapter-protocol/] support that
+Bloop offers. You can find the entire [debugging reference
+here](https://scalacenter.github.io/bloop/docs/debugging-reference).
+
+## Bloop CLI
+
+Especially for Metals users, I highly recommend using Bloop CLI. Up above we
```

Comment:
Sounds good. I made sure to emphasize what I more meant was that if you have a pretty simple workflow of compile, test, compile, etc, then that's where it maybe makes the most sense to utilize Bloop CLI.

I've heard you mention before that you avoid it when using Metals. I still haven't ever hit on a deadlock, although, I've only started heavily using in the past few months. It's been quite some time before I just ran a test through sbt rather than Bloop CLI, and I've been really happy with it.